File: chat-full.py
import boto3
from botocore.config import Config
import os
import json
import time
import logging
import threading
import copy
from boto3.dynamodb.conditions import Attr,Key

log = logging.getLogger()
log.setLevel(logging.INFO)

table = boto3.resource('dynamodb').Table(os.environ['Connections_Table'])
translate_client = boto3.client('translate', region_name='us-east-1')
comprehend_client = boto3.client('comprehend')

management_api = ""

def ws_connect(connection_id, nickname = ""):
    log.info('New connection established: {}'.format(connection_id))
    item = {'connectionId': connection_id}
    if len(nickname) != 0:
        item['nickname'] = nickname
    broadcast_message(connection_id, {'nickname': nickname, 'message': "Joined the chatroom"})
    response = table.put_item(Item = item)
    return {'statusCode': response['ResponseMetadata']['HTTPStatusCode']}
    

def ws_disconnect(connection_id):
    log.info('Deleting connection {}'.format(connection_id))
    nicknames = table.scan(
        ProjectionExpression='nickname',
        FilterExpression=Key("connectionId").eq(connection_id)
    )
    response = table.delete_item(
        Key={
            'connectionId': connection_id
        }
    )
    if len(nicknames[u'Items'])!=0:
        nickname = nicknames[u'Items'][0]['nickname']
    else:
        nickname = ""
    broadcast_message(connection_id, {'nickname': nickname, 'message': "left the chatroom"})
    return response['ResponseMetadata']

def broadcast_message(current_connectId, post_data):
    connections = table.scan(
        ProjectionExpression='connectionId'
    )
    threadList = []
    for item in connections['Items']:
        connection_id = item['connectionId']
        messagepayload = copy.deepcopy(post_data);
        if connection_id == current_connectId:
            messagepayload['nickname'] = 'You'
        t = threading.Thread(target=send_single_message, args=(connection_id, messagepayload))
        t.setDaemon(True)
        t.start()
        threadList.append(t)
    for thread in threadList:
        thread.join()


def send_single_message(connection_id,  messagepayload):
    messagepayload['timestamp'] = time.time();
    try:
        management_api.post_to_connection(
            ConnectionId=connection_id,
            Data=json.dumps(messagepayload)
        )
    except Exception as e:
        log.info('Error posting data: {}'.format(e.response['Error']))
        if e.response['Error']['Code'] == '410':
            log.info('Found stale connection')
            ws_disconnect(connection_id)


def ws_message(eventRequestContext, eventBody):
    body = json.loads(eventBody)
    post_data = {
        'message': body['message']
    }

    if body['translate'] == True:
        try:
            translate_response = translate_client.translate_text(
                Text=body['message'],
                SourceLanguageCode='auto',
                TargetLanguageCode='en'
            )
            if translate_response['ResponseMetadata']['HTTPStatusCode'] == 200:
                post_data['message'] = translate_response['TranslatedText']
        except Exception as e:
            log.info('Can not translate the input')

    if body['sentiment'] == True:
        comprehend_response = comprehend_client.detect_sentiment(
            LanguageCode='en',
            Text=post_data['message']
        )
        post_data['sentiment'] = comprehend_response['Sentiment']

    post_data['nickname'] = body.get('nickname','')

    if body['message'].startswith('@'):
        to_name = body['message'].split(" ")[0][1:]
        post_data['message'] = " ".join(post_data['message'].split(" ")[1:])
        connections = table.scan(
            ProjectionExpression='connectionId',
            FilterExpression=Attr("nickname").eq(to_name)
        )
        for connection in connections[u'Items']:
            send_single_message(connection[u'connectionId'], post_data)
    else:
        broadcast_message(eventRequestContext['connectionId'], post_data)
    return {'statusCode': 200}


def lambda_handler(event, context):
    print(event)
    global management_api
    if management_api == "":
        endpoint_url = 'https://{}/{}'.format(event['requestContext']['domainName'], event['requestContext']['stage'])
        config = Config(
            retries={'max_attempts': 1},
            read_timeout=5,
            connect_timeout=5
        )
        management_api = boto3.client('apigatewaymanagementapi', endpoint_url=endpoint_url, config=config)
    if event['requestContext']['eventType'] == 'CONNECT':
        if 'queryStringParameters' in event:
            nickname = event['queryStringParameters']['nickname']
        else:
            nickname = ""
        response = ws_connect(event['requestContext']['connectionId'], nickname)
        return response

    if event['requestContext']['eventType'] == 'DISCONNECT':
        response = ws_disconnect(event['requestContext']['connectionId'])
        return {'statusCode': response['HTTPStatusCode']}

    if event['requestContext']['eventType'] == 'MESSAGE':
        response = ws_message(event['requestContext'], event['body'])
        return response
    #We should never reach here
    return {'statusCode': 404}

File: chat-full.js
var AWS = require('aws-sdk');
AWS.config.update({ region: process.env.AWS_REGION });
var DDB = new AWS.DynamoDB({ apiVersion: "2012-10-08" });
const tablename = process.env.TABLE_NAME;
var translate = new AWS.Translate({ apiVersion: '2017-07-01' });
var comprehend = new AWS.Comprehend();

require('aws-sdk/clients/apigatewaymanagementapi');

var apigwManagementApi;

async function wsconnect(connectionId, nickname) {
  var putParams = {
    TableName: tablename,
    Item: {
      connectionId: { S: connectionId }
    }
  };
  if (nickname) {
    putParams.Item.nickname = { S: nickname }
  }
  await send_messages("", { "message": "Joined the chatroom", "nickname": nickname });
  await DDB.putItem(putParams).promise();
}

async function ws_disconnect(connection_id) {
  var scanParams = {
    FilterExpression: "connectionId = :a",
    ProjectionExpression: "nickname",
    TableName: tablename,
    ExpressionAttributeValues: {
      ":a": {
        S: connection_id
      }
    }
  };
  var nicknames = await DDB.scan(scanParams).promise();
  var deleteParams = {
    TableName: tablename,
    Key: {
      connectionId: { S: connection_id }
    }
  };
  await DDB.deleteItem(deleteParams).promise();
  console.log(JSON.stringify(nicknames));
  const nickname = (nicknames.Items[0].hasOwnProperty('nickname')) ? nicknames.Items[0].nickname.S : "";

  await send_messages(connection_id, { 'nickname': nickname, 'message': "left the chatroom" });
}

async function send_messages(current_connection, message, to_name) {
  message.timestamp = Date.now()/1000;
  var scanParams = {
    TableName: process.env.TABLE_NAME,
    ProjectionExpression: "connectionId"
  };

  if (to_name) {
    scanParams.FilterExpression = "nickname = :a",
      scanParams.ExpressionAttributeValues = {
        ":a": {
          S: to_name
        }
      }
  }

  var connections = await DDB.scan(scanParams).promise();
  var promises = [];
  connections.Items.forEach(function (element) {
    promises.push(new Promise(function (resolve, reject) {
      var sentMessage = JSON.parse(JSON.stringify(message));
      if (element.connectionId.S == current_connection) {
        sentMessage.nickname = "You";
      }
      var post_data = {
        Data: JSON.stringify(sentMessage),
        ConnectionId: element.connectionId.S
      };
      apigwManagementApi.postToConnection(post_data, (err, data) => {
        if (err) {
          ws_disconnect(element.connectionId.S);
          reject(err);
        } else {
          resolve(data);
        }
      });
    }));
  });
  await Promise.all(promises).then(function (values) { }).catch(function (values) {
    console.log(values);
  });
}

async function ws_message(connection_id, eventBody) {
  var eventBodyJson = JSON.parse(eventBody);

  if (eventBodyJson.translate) {
    var translateParams = {
      SourceLanguageCode: 'auto', /* required */
      TargetLanguageCode: 'en', /* required */
      Text: eventBodyJson.message
    };
    try {
      var translatedata = await translate.translateText(translateParams).promise();
      eventBodyJson.message = translatedata.TranslatedText;
    }
    catch (err) {
      console.log("translate failed...still use the original one");
    }
  }

  var post_data = {
    'message': eventBodyJson.message,
    'nickname': eventBodyJson.nickname
  }

  if (eventBodyJson.sentiment) {
    var params = {
      LanguageCode: 'en',
      Text: eventBodyJson.message
    };
    try {
      var comprehenddata = await comprehend.detectSentiment(params).promise();
      post_data.sentiment = comprehenddata.Sentiment;
    }
    catch (err) {
      console.log("detectSentiment failed...still use the original one");
    }
  }
  
  if (post_data.message.startsWith("@")) {
    var splitWords = post_data.message.split(" ");
    const to_name = splitWords.shift();
    post_data.message = splitWords.join(" ");
    await send_messages(connection_id, post_data, to_name.slice(1, to_name.length))
  } else {
    await send_messages(connection_id, post_data);
  }
}

exports.handler = async (event) => {
  if (!apigwManagementApi) {
    apigwManagementApi = new AWS.ApiGatewayManagementApi({
      apiVersion: "2018-11-29",
      endpoint: event.requestContext.domainName + "/" + event.requestContext.stage
    });
  }

  if (event['requestContext']['eventType'] == 'CONNECT') {
    var nickname = (event.hasOwnProperty("queryStringParameters")) ? event['queryStringParameters']['nickname'] : ""
    await wsconnect(event.requestContext.connectionId, nickname);
    return {
      statusCode: 200,
      body: "Connected."
    };
  }

  if (event['requestContext']['eventType'] == 'DISCONNECT') {
    return ws_disconnect(event.requestContext.connectionId);
  }

  if (event['requestContext']['eventType'] == 'MESSAGE') {
    await ws_message(event.requestContext.connectionId, event.body);
    return {
      statusCode: 200,
      body: "message sent."
    };
  }
};

